lib/package.gi

```diff
@@ -1190,30 +1189,27 @@ InstallGlobalFunction( DefaultPackageBannerString, function( inforec )
 #F  DirectoriesPackagePrograms( <name> )
 ##
 InstallGlobalFunction( DirectoriesPackagePrograms, function( name )
-    local arch, dirs, info, version, r, path;
+    local info, ppath;
 
-    arch := GAPInfo.Architecture;
-    dirs := [];
     # We are not allowed to call
     # `InstalledPackageVersion', `TestPackageAvailability' etc.
     info:= PackageInfo( name );
     if IsBound( GAPInfo.PackagesLoaded.( name ) ) then
       # The package is already loaded.
-      version:= GAPInfo.PackagesLoaded.( name )[2];
+      ppath:= GAPInfo.PackagesLoaded.( name )[1];
     elif IsBound( GAPInfo.PackageCurrent ) then
       # The package is currently going to be loaded.
-      version:= GAPInfo.PackageCurrent.Version;
+      ppath:= GAPInfo.PackageCurrent.InstallationPath;
     elif 0 < Length( info ) then
-      # Take the installed package with the highest version.
-      version:= info[1].Version;
+      # Take the installed package with the highest version
+      # that has been found first in the root paths.
+      ppath:= info[1].InstallationPath;
```

Comment:
So, that's a semantical change now, isn't it? Before we took all packages with the same version, and for each took its path. Now, you only take the first one, and ignore the rest.

Is that intentional? Why? Also, is this covered by the PR description? Maybe it is, but I don't quite understand all in there, either. E.g. what does this mean:
> DirectoriesPackagePrograms and DirectoriesPackageLibrary now search according to the installation path, not to the version number

I think this confused me because "search" is ambiguous here. When reading that comment, I am thinking about loading packages, and then "search" immediately leads me to think about "searching the pkg dirs in the GAP roots for PackageInfo.g files".

Anyway, I really had to look at the code to understand what is going on, and even now, I am not sure I fully understand. I need to look at that code in a wider context, I think (i.e., open up the file, and read)

Reply:
I do not understand the comment.
The list `info` contains all info records for the package in question, sorted by decreasing version number (and in the new version, where several records for the same version can exist, also sorted according to the ordering of the root paths where they were found).
In the old version, we fetch the version number and later look up the (unique) info record with this version number in order to fetch its path.
In the new version, we fetch the path directly.
(Perhaps the reason for the confusion is the `for` loop in the old version, which did not find more than one entry to add because there was only one package for the given version.)

Reply:
Yes, the confusion is: *why* is there only one record for each given version? My (apparently wrong) impression was that the plan with this patch was to add *all* copies of a given package we find to our records, even those with identical version; but keep them sorted in the order we find them; and then when looking for binaries, we search all copies of the latest version. But it seems I misunderstood the intentions here?

And I still think that the phrase "search according to the installation path" is completely unclear. You now wrote "sorted according to the ordering of the root paths where they were found" which makes much more senes to me.

Reply:
@fingolfin Thanks for the explanations.
As I had written in the description, the idea is to admit instances of a package with the same version numbers, in the sense that the package is regarded as loadable if there is one instance which is loadable.
This means in particular that it is not supported to "distribute the parts of a package to different root directories", for example to have a valid package binary only in one directory, some GAP functions of the package only in a second directory, and some package data only in a third directory.

(Concerning the test failure due to decreased coverage:
The coverage decreases when one removes some useless but covered code lines.
And when one changes or moves an uncovered line then apparently this counts as the introduction of a new uncovered line.
Very helpful.)

Reply:
The coverage failure does not block merging, only actual test suite failures do (this is configured in the "branch protection rules"). The PR status says "Merging is blocked" not because of the "failing" coverage, but because there is no approving review yet.